os/lib/ccm-star.c

```diff
@@ -42,64 +42,68 @@
 #include "lib/aes-128.h"
 #include <string.h>
 
-/* see RFC 3610 */
-#define CCM_STAR_AUTH_FLAGS(Adata, M) ((Adata ? (1u << 6) : 0) | (((M - 2u) >> 1) << 3) | 1u)
+/* As per RFC 3610. M == 2 (m_len is two bytes long) */
+#define CCM_STAR_AUTH_FLAGS(AUTH, MICLEN) (((AUTH) ? (1u << 6) : 0) | ((((MICLEN) - 2u) >> 1) << 3) | 1u)
 #define CCM_STAR_ENCRYPTION_FLAGS     1
+/* The auth data can have a variable length, but this implementation supports
+ * only the case of 0 < l(a) < (2^16 - 2^8) */
+#define MAX_A_LEN 0xfeff
 
 /*---------------------------------------------------------------------------*/
 static void
 set_iv(uint8_t *iv,
     uint8_t flags,
     const uint8_t *nonce,
-    uint8_t counter)
+    uint16_t counter)
 {
   iv[0] = flags;
   memcpy(iv + 1, nonce, CCM_STAR_NONCE_LENGTH);
-  iv[14] = 0;
+  iv[14] = counter >> 8;
   iv[15] = counter;
 }
 /*---------------------------------------------------------------------------*/
 /* XORs the block m[pos] ... m[pos + 15] with K_{counter} */
 static void
 ctr_step(const uint8_t *nonce,
-    uint8_t pos,
+    uint16_t pos,
     uint8_t *m_and_result,
-    uint8_t m_len,
-    uint8_t counter)
+    uint16_t m_len,
+    uint16_t counter)
 {
   uint8_t a[AES_128_BLOCK_SIZE];
   uint8_t i;
-  
+
   set_iv(a, CCM_STAR_ENCRYPTION_FLAGS, nonce, counter);
   AES_128.encrypt(a);
-  
+
   for(i = 0; (pos + i < m_len) && (i < AES_128_BLOCK_SIZE); i++) {
     m_and_result[pos + i] ^= a[i];
   }
 }
 /*---------------------------------------------------------------------------*/
 static void
 mic(const uint8_t *nonce,
-    const uint8_t *m, uint8_t m_len,
-    const uint8_t *a, uint8_t a_len,
+    const uint8_t *m, uint16_t m_len,
+    const uint8_t *a, uint16_t a_len,
     uint8_t *result,
     uint8_t mic_len)
 {
   uint8_t x[AES_128_BLOCK_SIZE];
-  uint8_t pos;
+  uint32_t pos; /* 32-bits as can need to exceed m_len to reach end of loop */
   uint8_t i;
-  
-  set_iv(x, CCM_STAR_AUTH_FLAGS(a_len, mic_len), nonce, m_len);
+
```

Comment:
Would it be nicer to check if the given `mic_len` is valid?
```
[Section 2, RFC 3610]

   For the generic CCM mode there are two parameter choices.  The first
   choice is M, the size of the authentication field.  .... (snip) ...
   Valid values are 4, 6, 8, 10, 12, 14, and 16 octets.  The second
```

Reply:
agree

Reply:
done